Fix #5578 namespaces/use placement

rossluk opened the pull request:
Fix "Fix Import" namespaces/uses placement in multiple php open tag files, when first line of php file is not empty or it is a block element, after declare statements in different scope

lbownik commented on php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java:
```diff
@@ -148,14 +146,24 @@ private void processSelections() {
+        }  catch (BadLocationException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+        return false;
```
would return in catch block read better?

rossluk replied:
i am not a java developer, if it would be better i am willing to make changes, but just checked few files with similar logic and there is usually return outside

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java:
```diff
@@ -148,14 +146,24 @@ private void processSelections() {
     }
+    
+    private boolean shouldAppendLineAfter(int lineOffset)
+    {
```
Formatting: Please move "{" to the end of the previous line.

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java:
```diff
@@ -527,7 +519,15 @@ public void visit(DeclareStatement node) {
+                // only - declare (strict_types=1) - should go before a namespace
+                // and such a declare can't be a block {} element or can't be in a global
+                // namespace, but - declare(ticks=1) could be anywhere.
+                if (i.getName().equals("strict_types")) {
```
Please add `// NOI18N`

junichi11 commented:
Could you add unit tests?

tmysik reviewed (changes requested):
Please, add unit tests for your use case. Thank you.

rossluk commented:
I've found more issues related to use placement, will do force push when have it fixed

rossluk commented:
Added new test case and refactored some code. There is still a problem with multiple global namespaces, but problem is not related to Fix Import and I suppose should be fixed by other PR.

junichi11 commented:
Please rename `testNB5578` to `testGH5578`.
Please add issue number to the commit message.

tmysik reviewed (approved):
LGTM, thanks for your work!

tmysik commented:
@rossluk Please, wait for approval from @junichi11. Thank you.

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java:
```diff
@@ -522,12 +544,29 @@ public List<NamespaceDeclaration> getGlobalNamespaceDeclarations() {
+
+        @Override
+        public void visit(UseStatement node) {
+            usedRanges.add(new OffsetRange(node.getStartOffset(), node.getEndOffset()));
```
Please add 
```java
            if (CancelSupport.getDefault().isCancelled()) {
                return;
            }
```

rossluk replied:
Done

junichi11 commented on php/php.editor/test/unit/data/testfiles/actions/testGH5578/01/testGH5578_01.php.fixUses:
```diff
@@ -0,0 +1,39 @@
+<?php 
+
+use NS1\B;
+declare (ticks=1) {
+    
+    
+}
```
We should add use statements below the declare statement if possible.
see: https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements

Please add a test case that has also `namespace` (`declare` and `namespace`).

rossluk replied:
Updated logic due to it, added new tests, also added a check for USE placing after phpDoc if there is no namespace in file (if namespace and phpDoc are there it will be placed after namespace).

junichi11 reviewed (approved):
Thank you for your work.
